### PR TITLE
Move `config.py` functionalities to the framework

### PR DESCRIPTION
**Pull request by gabor-mezei-arm**
Move the part of the `config.py` which is common in the TF-PSA_Crypto and the MbedTLS repo to the framework repo.

Related to Mbed-TLS/mbedtls#9470

### CODE REVIEW

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [inline scripts/mbedtls_framework/config_common.py:333]: About `file_type` I think we could rather just pass the configuration file default path, that's all we use from `file_type`.

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [inline scripts/mbedtls_framework/config_common.py:342]:
```suggestion
        self.args = self.parser.parse_args()
```
?

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [inline scripts/mbedtls_framework/config_common.py:398]: Using an `parser_args` or just `args` and `config` local variable would improve the readability to me:
```
args = self.parser_args
config = self.config
```

File: scripts/mbedtls_framework/config_common.py
Comment (gabor-mezei-arm) [reply scripts/mbedtls_framework/config_common.py:178]: Actually the `if name` is also checking if `name` is not empty string.

**Comment by gabor-mezei-arm** [comment]
During the fixes I found that the implementation and documentation of `get` function is differ. This function was not used before and was never changed the initial commit. Should I change the documentation or the implementation?

https://github.com/Mbed-TLS/mbedtls-framework/blob/025c8e7419164d6d2819141e034628b8f6d3eb13/scripts/mbedtls_framework/config_common.py#L85

**Comment by gilles-peskine-arm** [comment]
> I found that the implementation and documentation of get function is differ.

I don't see the problem, can you clarify?

> This function was not used before and was never changed after the initial commit.

When I wrote `config.py`, I expected that we'd use `get` often in Python scripts that depend on the configuration. But it turns out that we haven't written many scripts that depend on the configuration until now: we still mostly use `config.py` only through the command line.

> I think it is meant to be:
>
>         if name in self:
>             return self.settings[name].value

I don't see the difference: the `__contains__` method of `self` defines `… in self` to be equivalent to `… in self.settings`.

**Comment by gabor-mezei-arm** [comment]
> I don't see the difference: the `__contains__` method of `self` defines `… in self` to be equivalent to `… in self.settings`.

The difference is the `... in self`/`__contains__` returns true only when a setting is active (the macro is enabled) and the `... in self.settings` returns true if a setting is present (active or inactive). So for the inactive settings the `__contains__` returns false which is what the documentation wants.

**Comment by gilles-peskine-arm** [comment]
> The difference is the ... in self/__contains__ returns true only when a setting is active (the macro is enabled) and the ... in self.settings returns true if a setting is present (active or inactive).

Oh, right. Yes, I think the intent of `get` was to be the way it's documented, not the way it behaves.

It doesn't matter much since we aren't using that aspect, but please do fix it. It's always better when the code matches the documentation. Thanks for spotting that!

(This script should really have unit tests, but that's just one item in the huge tech debt pile.)

File: scripts/mbedtls_framework/config_common.py
Comment (gabor-mezei-arm) [reply scripts/mbedtls_framework/config_common.py:185]: I agree with there should be a list. The real question is do we want to differentiate the single and multiple config handler descendants? The latter is using more then one config file and need to specify all of them or none of them to work. Not to mention the order, it can be easily ending up with swapping the content of the files.

IMO the solution would be using str for single configs handlers and list of tupples or a dict for multiple config handlers.

File: scripts/mbedtls_framework/config_common.py
Comment (davidhorstmann-arm) [reply scripts/mbedtls_framework/config_common.py:330]: Hmm

I think `custom_parser_options` is fine - if you're only calling `parser.add_argument()` (which I believe is perfectly valid here) then you aren't really adding a parser per se, you're adding a new option to the existing parser.

However, we are missing a plural here:
```suggestion
    Custom parser options can be added by overriding 'custom_parser_options'.
```

**Review by davidhorstmann-arm — COMMENTED** [review]
LGTM (obviously pending Ronald being happy). I've verified that the move is correct (moved classes are identical to their former selves.

That being said, things I find when I complete the review of #9470 might need to be addressed here, so I'm holding back on a full approval.

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [reply scripts/mbedtls_framework/config_common.py:330]: Ah right, thanks. Thus yes, just `Custom parser options` instead of `Custom parser option` line 330.

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [reply scripts/mbedtls_framework/config_common.py:185]: Looking at this, it seems to be the case that writing a configuration associated to the crypto and mbedtls configuration file (class CombinedConfig) is actually working fine and that the only "issue" we have currently revealing that things could maybe be better shaped is the "#pylint: disable=arguments-differ" needed for the CombinedConfig when overwriting the write method?
If that's the case and thus the writing of the two configuration files with the CombinedConfig `write` method is working fine I am happy to leave things as they are. I would only ask for a comment in config.py explaining the "#pylint: disable=arguments-differ" mentioned above.

File: scripts/mbedtls_framework/config_common.py
Comment (ronald-cron-arm) [reply scripts/mbedtls_framework/config_common.py:172]: okay, thanks. Resolving this.

**Comment by davidhorstmann-arm** [comment]
I've left 2 comments, LGTM otherwise. This review is intertwined with that of the other PR (Mbed-TLS/mbedtls#9470) as well, so once this one is approved the other one should also be fine.

**Review by ronald-cron-arm — COMMENTED** [review]
Some documentation comments but otherwise this looks good to me but some doubts about https://github.com/Mbed-TLS/mbedtls-framework/pull/41#discussion_r1765080757.

File: scripts/mbedtls_framework/config_common.py
Comment (gilles-peskine-arm) [reply scripts/mbedtls_framework/config_common.py]: `:param` is the documentation syntax for Sphinx (a variant of reStructuredText). Sphinx is a documentation system that's very popular with Python, but [it's not the only one](https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats). Our docstrings mostly don't follow the Sphinx syntax. Unless we decide to adopt Sphinx, there's no point in using Sphinx syntax in a single function in a module.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thsnks.

**Comment by ronald-cron-arm** [comment]
The CI on #9470 and #9511 is green.
I've run test_config_script.py on the base and head of the two PRs and it does not detect any regression.
This validates this PR thus merging.

File: scripts/generate_config_tests.py
Comment (ronald-cron-arm) [inline scripts/generate_config_tests.py:165]: @gabor-mezei-arm
Now, in build_tree.py we have an is_mbedtls_3_6() method.
Is it the case that if we restore here something like:
```py
self.psa_config = config.ConfigFile('include/psa/crypto_config.h') if build_tree.is_mbedtls_3_6() else \
config.CryptoConfig.h
```
we could then remove CryptoConfig() and CryptoConfigFile() from config.py in mbedtls-3.6?

Not saying that we should do it, just checking my understanding.

Comment (gabor-mezei-arm) [reply scripts/generate_config_tests.py:165]: `ConfigFile` changed to an abstract class. It can not be instantiated.

**Pull request merged**
